src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Modules.java

```diff
@@ -536,7 +536,7 @@ private void setCompilationUnitModules(List<JCCompilationUnit> trees, Set<Module
     private void checkSourceLocation(JCCompilationUnit tree, ModuleSymbol msym) {
         try {
             JavaFileObject fo = tree.sourcefile;
-            if (fileManager.contains(msym.sourceLocation, fo)) {
+            if (msym.sourceLocation != null && fileManager.contains(msym.sourceLocation, fo)) {
```

Comment:
What is the module for which the `sourceLocation` is `null` ... in other words, it is reasonable for this method to be called with a module with no source location, or is this check just hiding some other bug elsewhere?

Reply:
@jonathan-gibbons Thanks for the comment. Your worry is right. I do more investigation in the source code today.

The logic of the method `ClassFinder#scanUserPaths` seems wrong. 
When the `fileManager.hasLocation(SOURCE_PATH)` (means `haveSourcePath` in my updated code) is true and the `includeSourcePath` is false, the compiler shouldn't search any source path and source files.

I seperated the situations and revised the code and test. What do you think about the updated code?
I worry that this change may cause some regressions. Currently, the `tools/javac` passed locally.
I am still waiting for the `Pre-submit tests` to finish.

Reply:
And about the null check, I want to keep it so that the compiler won't crash when meeting the situation like this bug.

Reply:
> And about the null check, I want to keep it so that the compiler won't crash when meeting the situation like this bug.

If we believe that the null should not occur, it would be covering up a bug to provide the null check.

Reply:
Got it. I can't find a situation now that the null occurs. So I removed the null check just now.